### ISSUE TITLE
Solaris11: check-iprop fails

### ISSUE THREAD

**Issue by quanah** [heimdal/heimdal#380]
After applying the fixes from @lhoward for the PKCS11 backend, I'm seeing failures with check-iprop:

```
quanah@sol11-build:~/git/sold/sold-master/SunOS5.11-i86pc/build/heimdal.i86pc_64/tests/kdc$ ./check-iprop
Creating database
Test log recovery
Starting kdc
starting master
starting slave
Waiting for ipropd-slave to start, looking logfile messages.log
Have waited 2 seconds
Have waited 4 seconds
Have waited 6 seconds
Have waited 8 seconds
Have waited 10 seconds
Have waited 12 seconds
Have waited 14 seconds
Have waited 16 seconds
Have waited 18 seconds
Have waited 20 seconds
Have waited 22 seconds
Have waited 24 seconds
Have waited 26 seconds
Have waited 28 seconds
Have waited 30 seconds
Have waited 32 seconds
Have waited 34 seconds
Have waited 36 seconds
Have waited 38 seconds
Have waited 40 seconds
Have waited 42 seconds
Have waited 44 seconds
Have waited 46 seconds
Have waited 48 seconds
Have waited 50 seconds
Have waited 52 seconds
Have waited 54 seconds
Have waited 56 seconds
Have waited 58 seconds
Have waited 60 seconds
Have waited 62 seconds
Have waited 64 seconds
Have waited 66 seconds
Waited for 65 for the ipropd-slave to start, and it didnt happen
killing ipropd s + m + kdc
2018-05-11T08:33:37 slave status change: connecting to master: localhost


2018-05-11T08:33:37 connection successful to master: localhost[127.0.0.1]
2018-05-11T08:33:37 Using nss for name canon immediately
2018-05-11T08:33:37 krb5_get_creds: requesting a ticket for iprop/localhost@TEST.H5L.SE
2018-05-11T08:33:37 sha1 checksum failed
2018-05-11T08:33:39 krb5_recvauth: End of file
2018-05-11T08:33:39 Child was killed; waiting 30 seconds to restart
2018-05-11T08:33:50 ipropd-master terminated
master down at 2018-05-11T08:33:50
```

The messages.log file has the following:

```
2018-05-11T08:39:30 error message: Did not find a plugin for ccache_ops: 2
2018-05-11T08:39:30 error message: open or stat ../../tests/kdc/kdc.conf: No such file or directory: 2
2018-05-11T08:39:30 slave status change: bootstrapping

2018-05-11T08:39:30 slave status change: getting credentials from keytab/database

2018-05-11T08:39:30 error message: failed to open ../../tests/kdc/mkey.file: No such file or directory: 2
2018-05-11T08:39:30 slave status change: creating log file

2018-05-11T08:39:30 Using nss for name canon immediately
2018-05-11T08:39:30 krb5_get_init_creds: loop 1
2018-05-11T08:39:30 KDC send 0 patypes
2018-05-11T08:39:30 error message: Did not find a plugin for send_to_kdc: 2
2018-05-11T08:39:30 Trying to find service kdc for realm TEST.H5L.SE flags 0
2018-05-11T08:39:30 error message: Did not find a plugin for service_locator: 2
2018-05-11T08:39:30 configuration file for realm TEST.H5L.SE found
2018-05-11T08:39:30 submissing new requests to new host
2018-05-11T08:39:30 error message: Did not find a plugin for send_to_kdc: 2
2018-05-11T08:39:30 connecting to host: udp 127.0.0.1:49188 (localhost) tid: 00000001
2018-05-11T08:39:30 writing packet: udp 127.0.0.1:49188 (localhost) tid: 00000001
2018-05-11T08:39:30 AS-REQ iprop/slave.test.h5l.se@TEST.H5L.SE from IPv4:127.0.0.1 for iprop/localhost@TEST.H5L.SE
2018-05-11T08:39:30 Client sent patypes: REQ-ENC-PA-REP
2018-05-11T08:39:30 Looking for PK-INIT(ietf) pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 Looking for PK-INIT(win2k) pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 Looking for ENC-TS pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 Need to use PA-ENC-TIMESTAMP/PA-PK-AS-REQ
2018-05-11T08:39:30 sending 297 bytes to IPv4:127.0.0.1
2018-05-11T08:39:30 reading packet: udp 127.0.0.1:49188 (localhost) tid: 00000001
2018-05-11T08:39:30 host completed: udp 127.0.0.1:49188 (localhost) tid: 00000001
2018-05-11T08:39:30 error message: Need to use PA-ENC-TIMESTAMP/PA-PK-AS-REQ: -1765328359
2018-05-11T08:39:30 unknown-function TEST.H5L.SE done: 0 hosts: 1 packets: 1 wc: 0.005120 nr: 0.000272 kh: 0.000137 tid: 00000001
2018-05-11T08:39:30 krb5_get_init_creds: loop 2
2018-05-11T08:39:30 krb5_get_init_creds: processing input
2018-05-11T08:39:30 krb5_get_init_creds: got an error
2018-05-11T08:39:30 error message: Need to use PA-ENC-TIMESTAMP/PA-PK-AS-REQ: -1765328359
2018-05-11T08:39:30 krb5_get_init_creds: KRB-ERROR -1765328359
2018-05-11T08:39:30 KDC send 7 patypes
2018-05-11T08:39:30 KDC send PA-DATA type: 16
2018-05-11T08:39:30 KDC send PA-DATA type: 15
2018-05-11T08:39:30 KDC send PA-DATA type: 147
2018-05-11T08:39:30 KDC send PA-DATA type: 2
2018-05-11T08:39:30 KDC send PA-DATA type: 138
2018-05-11T08:39:30 KDC send PA-DATA type: 136
2018-05-11T08:39:30 KDC send PA-DATA type: 19
2018-05-11T08:39:30 krb5_get_init_creds: using ENC-TS with enctype 18
2018-05-11T08:39:30 error message: Name canon failed while searching keytab: 0
2018-05-11T08:39:30 error message: Did not find a plugin for send_to_kdc: 2
2018-05-11T08:39:30 Trying to find service kdc for realm TEST.H5L.SE flags 0
2018-05-11T08:39:30 error message: Did not find a plugin for service_locator: 2
2018-05-11T08:39:30 configuration file for realm TEST.H5L.SE found
2018-05-11T08:39:30 submissing new requests to new host
2018-05-11T08:39:30 error message: Did not find a plugin for send_to_kdc: 2
2018-05-11T08:39:30 connecting to host: udp 127.0.0.1:49188 (localhost) tid: 00010001
2018-05-11T08:39:30 writing packet: udp 127.0.0.1:49188 (localhost) tid: 00010001
2018-05-11T08:39:30 AS-REQ iprop/slave.test.h5l.se@TEST.H5L.SE from IPv4:127.0.0.1 for iprop/localhost@TEST.H5L.SE
2018-05-11T08:39:30 Client sent patypes: ENC-TS, REQ-ENC-PA-REP
2018-05-11T08:39:30 Looking for PK-INIT(ietf) pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 Looking for PK-INIT(win2k) pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 Looking for ENC-TS pa-data -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 ENC-TS Pre-authentication succeeded -- iprop/slave.test.h5l.se@TEST.H5L.SE using aes256-cts-hmac-sha1-96
2018-05-11T08:39:30 ENC-TS pre-authentication succeeded -- iprop/slave.test.h5l.se@TEST.H5L.SE
2018-05-11T08:39:30 error message: encryption type des-cbc-crc is disabled: -1765328234
2018-05-11T08:39:30 AS-REQ authtime: 2018-05-11T08:39:30 starttime: unset endtime: 2018-05-12T08:39:30 renew till: unset
2018-05-11T08:39:30 Client supported enctypes: aes256-cts-hmac-sha1-96, des3-cbc-sha1, arcfour-hmac-md5, using aes256-cts-hmac-sha1-96/aes256-cts-hmac-sha1-96
2018-05-11T08:39:30 sending 701 bytes to IPv4:127.0.0.1
2018-05-11T08:39:30 reading packet: udp 127.0.0.1:49188 (localhost) tid: 00010001
2018-05-11T08:39:30 host completed: udp 127.0.0.1:49188 (localhost) tid: 00010001
2018-05-11T08:39:30 unknown-function TEST.H5L.SE done: 0 hosts: 1 packets: 1 wc: 0.005535 nr: 0.000034 kh: 0.000041 tid: 00010001
2018-05-11T08:39:30 krb5_get_init_creds: loop 3
2018-05-11T08:39:30 krb5_get_init_creds: processing input
2018-05-11T08:39:30 krb5_get_init_creds: using keyproc
2018-05-11T08:39:30 error message: Name canon failed while searching keytab: 0
2018-05-11T08:39:30 krb5_get_init_creds: extracting ticket
2018-05-11T08:39:30 processing enc-ap-rep
2018-05-11T08:39:30 enc-ap-rep: valid
2018-05-11T08:39:30 Using nss for name canon immediately
2018-05-11T08:39:30 slave status change: ipropd-slave started

2018-05-11T08:39:30 slave status change: connecting to master: localhost


2018-05-11T08:39:30 connection successful to master: localhost[127.0.0.1]
2018-05-11T08:39:30 Using nss for name canon immediately
2018-05-11T08:39:30 krb5_get_creds: requesting a ticket for iprop/localhost@TEST.H5L.SE
2018-05-11T08:39:30 sha1 checksum failed
2018-05-11T08:39:31 krb5_recvauth: End of file
2018-05-11T08:39:31 Child was killed; waiting 30 seconds to restart
2018-05-11T08:40:01 Missed a signal, updating slaves 0 to 33
2018-05-11T08:40:01 slave status change: connecting to master: localhost


2018-05-11T08:40:01 connection successful to master: localhost[127.0.0.1]
2018-05-11T08:40:01 Using nss for name canon immediately
2018-05-11T08:40:01 krb5_get_creds: requesting a ticket for iprop/localhost@TEST.H5L.SE
2018-05-11T08:40:01 sha1 checksum failed
2018-05-11T08:40:02 krb5_recvauth: End of file
2018-05-11T08:40:02 Child was killed; waiting 30 seconds to restart
2018-05-11T08:40:32 slave status change: connecting to master: localhost


2018-05-11T08:40:32 connection successful to master: localhost[127.0.0.1]
2018-05-11T08:40:32 Using nss for name canon immediately
2018-05-11T08:40:32 krb5_get_creds: requesting a ticket for iprop/localhost@TEST.H5L.SE
2018-05-11T08:40:32 sha1 checksum failed
2018-05-11T08:40:32 krb5_recvauth: End of file
2018-05-11T08:40:32 Child was killed; waiting 30 seconds to restart
2018-05-11T08:40:40 ipropd-master terminated
```

**Comment by lhoward** [comment]
The issue is related to interaction between PKCS#11 and forking. I am investigating the correct fix for this.

**Comment by lhoward** [comment]
Essentially, PKCS#11 library state is invalid after fork() and C_Initialize() must be called again. The simplest fix for this is just to call C_Initialize() on every call. This has a performance impact (10ms on aes-test). Wrapping C_Initialize() in a separate pthread_once control (independent to PKCS#11 module load, so it could be invalidated in a pthread_atfork child handler) has about the same performance impact.

**Comment by lhoward** [comment]
Fixed in 934d5e09.
